src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+//  Copyright 2023, University of Freiburg,
+//                  Chair of Algorithms and Data Structures.
+//  Author: Noah Nock <noah.v.nock@gmail.com>
+#include <util/Rtree.h>
+#include <util/RtreeFileReader.h>
+#include <util/BackgroundStxxlSorter.h>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/serialization/vector.hpp>
+
+static void centerOrdering(multiBoxGeo& boxes, size_t dim) {
+  if (dim == 0) {
+    // order by centerX
+    SortRuleLambdaX comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  } else {
+    // order by centerY
+    auto sortRuleLambda = [](RTreeValue b1, RTreeValue b2) -> bool {
+      double center1 = std::midpoint(b1.box.min_corner().get<1>(), b1.box.max_corner().get<1>());
+      double center2 = std::midpoint(b2.box.min_corner().get<1>(), b2.box.max_corner().get<1>());
+      return center1 < center2;
+    };
+
+    std::sort(boxes.begin(), boxes.end(), sortRuleLambda);
+  }
+}
+
+static void centerOrdering(multiBoxWithOrderIndex& boxes, size_t dim) {
+  if (dim == 0) {
+    // order by centerX
+    SortRuleLambdaXWithIndex comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  } else {
+    // order by centerY
+    SortRuleLambdaYWithIndex comp;
+
+    std::sort(boxes.begin(), boxes.end(), comp);
+  }
+}
+
+OrderedBoxes SortInput(const std::string& onDiskBase, size_t M, uintmax_t maxBuildingRamUsage, bool workInRam) {
+  OrderedBoxes orderedInputRectangles;
+
+  ad_utility::BackgroundStxxlSorter<RTreeValue, SortRuleLambdaX> sorterRectsD0Basic = ad_utility::BackgroundStxxlSorter<RTreeValue, SortRuleLambdaX>(std::ceil((maxBuildingRamUsage < 9999999999 ? maxBuildingRamUsage : 9999999999) / 3.0));
+  multiBoxGeo rectsD0Basic;
+
+  if (workInRam) {
+    rectsD0Basic = Rtree::LoadEntries(onDiskBase + ".boundingbox.tmp");
+    centerOrdering(rectsD0Basic, 0);
+  } else {
+    FileReaderWithoutIndex fileReaderRectsD0 = FileReaderWithoutIndex(onDiskBase + ".boundingbox.tmp");
+    std::optional<RTreeValue> rectD0Element = fileReaderRectsD0.GetNextElement();
+    while (rectD0Element) {
+      sorterRectsD0Basic.push(rectD0Element.value());
+      rectD0Element = fileReaderRectsD0.GetNextElement();
+    }
+    fileReaderRectsD0.Close();
```

Comment:
```suggestion
   for (auto boundingBox: FileReaderWithoutIndex(filename)) {
     sorter.push(boundingBox);
   }
```
Of course the `FileReader` needs an interface that supports this, have a look at `cppcoro::generator` in `util/Generator.h` and its usages across the codebase.